artiq/firmware/runtime/moninj.rs

```diff
@@ -18,23 +18,23 @@ mod local_moninj {
             csr::rtio_moninj::mon_chan_sel_write(channel as _);
             csr::rtio_moninj::mon_probe_sel_write(probe);
             csr::rtio_moninj::mon_value_update_write(1);
-            csr::rtio_moninj::mon_value_read() as u32
+            csr::rtio_moninj::mon_value_read()
         }
     }
 
-    pub fn inject(channel: u16, overrd: u8, value: u8) {
+    pub fn inject(channel: u16, overrd: u8, value: u32) {
         unsafe {
             csr::rtio_moninj::inj_chan_sel_write(channel as _);
             csr::rtio_moninj::inj_override_sel_write(overrd);
-            csr::rtio_moninj::inj_value_write(value);
+            csr::rtio_moninj::inj_value_write(value.to_be());
         }
     }
 
-    pub fn read_injection_status(channel: u16, overrd: u8) -> u8 {
+    pub fn read_injection_status(channel: u16, overrd: u8) -> u32 {
```

Comment:
How does this work for the 48-bit FTW on AD9912?

Reply:
IIRC UrukulMonitor already implemented it. It split the data into two 32-bits of upper part and lower part. Upper is sent first then lower.

https://github.com/stevefan1999/artiq/blob/a14c8a46b36d1f401d395f4e605d224786f408ab/artiq/gateware/rtio/phy/urukul_monitor.py#L9-L11

So for 48-bit FTW it will be sent as a 32-bit upper and 16-bit lower chunk, where the top 16-bit in lower should be 0. It therefore has to be reconstructed. Summing the two parts up using a 64-bit int should do the trick. Then we perform FTW conversion using the sum